### PR TITLE
Added ok if let lint

### LINKED ISSUE #1185: Lint suggestion: Calling .ok() unnecessary in if let
I just found myself committing this code:

```rust
        for result in iter {
            if let Some(bench) = try!(result).parse().ok() {
                vec.push(bench)
            }
        }
```

I felt silly, because the `.ok()` is unnecessary, I could have just matched on `Ok(bench)`:

```rust
        for result in iter {
            if let Ok(bench) = try!(result).parse() {
                vec.push(bench)
            }
        }
```

So I figured this might be a good lint.

### PR DESCRIPTION
**Pull request by KitFreddura**
Hey all,

I added a lint that checks for expressions like:

```rust
if let Some(y) = x.parse().ok()
```
And suggests changing to:

```rust
if let Ok(y) = x.parse()
```

This is my first clippy lint, please let me know if I can refactor anything or make it more idiomatic!

Fix #1185.

### CODE REVIEW

File: clippy_lints/src/ok_if_let.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
+use rustc::lint::*;
+use rustc::hir::*;
+use utils::{method_chain_args, span_help_and_lint};
+/// **What it does:*** Checks for unnecessary ok() in if let.
```
Comment (mcarton) [inline clippy_lints/src/ok_if_let.rs]: We usually add a blank line here

File: clippy_lints/src/ok_if_let.rs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+///
+/// **Why is this bad?** Calling ok() in if let is unnecessary, instead match on Ok()
```
Comment (mcarton) [inline clippy_lints/src/ok_if_let.rs]: All code in ``` `` ```

File: clippy_lints/src/ok_if_let.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+/// for result in iter {
+///     if let Some(bench) = try!(result).parse().ok() {
+///         vec.push(bench)
+///     }
+/// }
+/// ```
+declare_lint! {
+    pub OK_IF_LET,
+    Warn,
+    "usage of ok() in if let statements is unnecessary, match on Ok(expr) instead"
```
Comment (mcarton) [inline clippy_lints/src/ok_if_let.rs]: Same here

File: clippy_lints/src/ok_if_let.rs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            let ExprMatch(ref op, ref body, ref source) = expr.node, //test if expr is a match
+            let MatchSource::IfLetDesugar { contains_else_clause: _ } = *source, //test if it is an If Let
+            let PatKind::TupleStruct(ref x, _, _)  = body[0].pats[0].node, //get operation
+            let Some(_) = method_chain_args(op, &["ok"]) //test to see if using ok() method
+
+        ], { 
+            if print::path_to_string(x) == "Some" { //if using ok() on a Some, kick in lint
+                span_help_and_lint(cx, OK_IF_LET, expr.span,
+                "Matching on `Some` with `ok()` is redundant",
+                "Consider matching on `Ok()` instead");
```
Comment (mcarton) [inline clippy_lints/src/ok_if_let.rs]: If you feel like it you can try to make an actual suggestion here.
(search for `span_suggestion`)

File: clippy_lints/src/ok_if_let.rs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    }
+}
+
+impl LateLintPass for OkIfLetPass {
+    fn check_expr(&mut self, cx: &LateContext, expr: &Expr) {
+        if_let_chain! {[
+            let ExprMatch(ref op, ref body, ref source) = expr.node, //test if expr is a match
+            let MatchSource::IfLetDesugar { contains_else_clause: _ } = *source, //test if it is an If Let
+            let PatKind::TupleStruct(ref x, _, _)  = body[0].pats[0].node, //get operation
+            let Some(_) = method_chain_args(op, &["ok"]) //test to see if using ok() method
```
Comment (mcarton) [inline clippy_lints/src/ok_if_let.rs]: What if my non-`Result` type happens to have an `ok` method? You should at least check the type of the expression is `Result` here.

**Comment by mcarton** [comment]
BTW, Travis fails because you need to run `./util/update_lints.py`.
Overall this looks like a good first try modulo the missing type-checking and other small stuffs.

File: clippy_lints/src/lib.rs [hunk 6]
```diff
@@ -403,6 +405,7 @@ pub fn register_plugins(reg: &mut rustc_plugin::Registry) {
         no_effect::NO_EFFECT,
         no_effect::UNNECESSARY_OPERATION,
         non_expressive_names::MANY_SINGLE_CHAR_NAMES,
+        ok_if_let::OK_IF_LET,
```
Comment (mcarton) [inline clippy_lints/src/lib.rs]: I don't really like that name. How about `IF_LET_SOME_RESULT`?

**Comment by KitFreddura** [comment]
Ok all that sounds good! I'm working on the updates right now, is there a way to typecheck an Expr? I'm looking for methods to ensure that the expr in question is of type Result.

**Comment by KitFreddura** [comment]
Updated as per requests:  
  
Added type checking to ensure that this lint is only occurring for expressions of the form Result<T,E>.ok() and added basic suggestion format.

**Comment by KitFreddura** [comment]
I'm a bit confused as to why the Travis build is failing, it looks like this lint is keeping clippy from compiling.

**Comment by mcarton** [comment]
Did you run *rustfmt* in 56a07d4? It did a lot of bad things…

File: clippy_lints/src/ok_if_let.rs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        lint_array!(IF_LET_SOME_RESULT)
+    }
+}
+
+impl LateLintPass for OkIfLetPass {
+    fn check_expr(&mut self, cx: &LateContext, expr: &Expr) {
+        if_let_chain! {[
+            let ExprMatch(ref op, ref body, ref source) = expr.node, //test if expr is a match
+            let MatchSource::IfLetDesugar { .. } = *source, //test if it is an If Let
+            let ExprMethodCall(_, _, ref result_types) = op.clone().unwrap().node, //get expr from expr.ok()
```
Comment (mcarton) [inline clippy_lints/src/ok_if_let.rs]: We never need to clone nodes. Here you can remove `clone().unwrap().`.

**Comment by KitFreddura** [comment]
Whoops, I forgot I had ```rustfmt``` on every save. I fixed the format issues and removed the ```clone().unwrap()```

**Comment by KitFreddura** [comment]
Had to refork the repo, creating new pull request elsewhere. Sorry for the confusion.

**Pull request closed without merging**